res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,746 @@
+            sendShifted: true,
+            shiftOffset: 0x08,
+            shift: function() {
+                this.inKey = "beatjump_" + this.size + "backward";
+                this.outKey = "beatjump_" + this.size + "backward";
+            },
+            unshift: function() {
+                this.inKey = "beatjump_" + this.size + "forward";
+                this.outKey = "beatjump_" + this.size + "forward";
+            },
+            outConnect: false
+        });
+    }
+};
+MixtrackProFX.ModeBeatjump.prototype = Object.create(components.ComponentContainer.prototype);
+
+MixtrackProFX.Browse = function() {
+    this.knob = new components.Encoder({
+        shiftControl: true,
+        shiftOffset: 0x01,
+        input: function(channel, control, value) {
+            var direction;
+            if (!MixtrackProFX.shifted) {
+                direction = (value > 0x40) ? value - 0x80 : value;
+                engine.setParameter("[Library]", "MoveVertical", direction);
+            } else {
+                direction = (value > 0x40) ? "up" : "down";
+                engine.setParameter("[Channel1]", "waveform_zoom_" + direction, 1);
+
+                // need to zoom both channels if waveform sync is disabled in Mixxx settings.
+                // and when it's enabled then no need to zoom 2nd channel, as it will cause
+                // the zoom to jump 2 levels at once
+                if (!MixtrackProFX.waveformsSynced) {
+                    engine.setParameter("[Channel2]", "waveform_zoom_" + direction, 1);
+                }
+            }
+        }
+    });
+
+    this.knobButton = new components.Button({
+        group: "[Library]",
+        inKey: "MoveFocusForward",
+        shiftControl: true,
+        shiftOffset: 0x01,
+        shift: function() {
+            this.inKey = "GoToItem";
+        },
+        unshift: function() {
+            this.inKey = "MoveFocusForward";
+        }
+    });
+};
+MixtrackProFX.Browse.prototype = new components.ComponentContainer();
+
+MixtrackProFX.Gains = function() {
+    this.mainGain = new components.Pot({
+        group: "[Master]",
+        inKey: "gain"
+    });
+
+    this.cueGain = new components.Pot({
+        group: "[Master]",
+        inKey: "headGain"
+    });
+
+    this.cueMix = new components.Pot({
+        group: "[Master]",
+        inKey: "headMix"
+    });
+};
+MixtrackProFX.Gains.prototype = new components.ComponentContainer();
+
+MixtrackProFX.vuCallback = function(value, group) {
+    var level = value * 90;
+    var deckOffset = script.deckFromGroup(group) - 1;
+    midi.sendShortMsg(0xB0 + deckOffset, 0x1F, level);
+};
+
+MixtrackProFX.scratchToggle = function(channel, control, value) {
+    if (value !== 0x7F) {
+        return;
+    }
+    MixtrackProFX.scratchModeEnabled[channel] = !MixtrackProFX.scratchModeEnabled[channel];
+    midi.sendShortMsg(0x90 | channel, 0x07, MixtrackProFX.scratchModeEnabled[channel] ? 0x7F : 0x01);
+};
```

Comment:
The entire scratch code is conceptually a per deck feature, so it would make to implement as part of that.